inspirehep/modules/disambiguation/core/ml/sampling.py

```diff
@@ -113,12 +114,14 @@ def sample_signature_pairs(signatures_path, clusters_path, pairs_size):
     # 3. Sample
     #
 
-    for category in [
-        same_cluster_same_name,
-        same_cluster_different_name,
-        different_cluster_same_name,
-        different_cluster_different_name,
-    ]:
-        if category:
-            for pair in np.random.choice(category, replace=True, size=(pairs_size // 4)):
-                yield pair
+    non_empty_categories = [
+        category for category in [
+            same_cluster_same_name,
+            same_cluster_different_name,
+            different_cluster_same_name,
+            different_cluster_different_name,
+        ] if category
+    ]
+    for _, category in itertools.izip(xrange(pairs_size), itertools.cycle(non_empty_categories)):
+        for pair in np.random.choice(category, size=1):
+            yield pair
```

Comment:
why not simply?
```python
for category in non_empty_categories:
    for pair in np.random.choice(category, replace=True, size=(pairs_size // len(non_empty_categories))):
        yield pair
```
This seems reasonable, but less efficient (`pairs_size` calls to the `np.random.choice` function vs `4`) and might be a bit harder to understand.

Reply:
That's an excellent idea, although it requires `pairs_size` to be a multiple of 12 (the LCM of 1, 2, 3, and 4, the number of possible non-empty categories) to make sure that we will always have that many samples.

Reply:
This happened in https://github.com/inspirehep/inspire-next/pull/3455/commits/62548f927675f0fd357d00612a42caf12b8790ed.